File: Solutions/Solution200.java
class Solution200 {
  public int numIslands(char[][] grid) {
    // Go through the whole grid and when we find a 1 we do a DFS
    // and change all the connected to 3. If we see a three or a 0
    // we ignore it, we count the number of times we find a 1 and
    // start a DFS

    if (grid.length == 1 && grid[0].length == 1) return (grid[0][0] == '1') ? 1 : 0;

    int islands = 0;

    for (int row = 0; row < grid.length; row++) {
      for (int col = 0; col < grid[0].length; col++) {
        if (grid[row][col] == '1') {
          islands++;
          System.out.println("Row: " + row + ", Col: " + col);
          islandDFS(grid, row, col);
        }
      }
    }

    return islands;
  }

  private void islandDFS(char[][] grid, int row, int col) {
    if (row < 0 || row > grid.length - 1 || col < 0 || col > grid[0].length - 1) return;

    if (grid[row][col] == '1') grid[row][col] = '0';

    if (row > 0 && grid[row - 1][col] == '1') islandDFS(grid, row - 1, col);
    if (row < grid.length - 1 && grid[row + 1][col] == '1') islandDFS(grid, row + 1, col);
    if (col > 0 && grid[row][col - 1] == '1') islandDFS(grid, row, col - 1);
    if (col < grid[0].length - 1 && grid[row][col + 1] == '1') islandDFS(grid, row, col + 1);
  }
}

File: Solutions/Solution200.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        
        # BFS and count distinct
        
        # Find all the 1s
        starting_points = []
        
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1":
                    starting_points.append((row, col))

        print(starting_points)
        
        distinct = 0
        seen = set()

        for coord in starting_points:
            
            if coord in seen:
                continue
                
            distinct += 1
            
            # BFS
            queue = deque()
            queue.append(coord)
            
            while queue:
                current = queue.popleft()
                
                if (current in seen or
                    current[0] < 0 or
                    current[0] >= len(grid) or
                    current[1] < 0 or
                    current[1] >= len(grid[0]) or
                    grid[current[0]][current[1]] != "1"):
                    continue
                
                seen.add(current)
                
                queue.append((current[0] + 1, current[1]))
                queue.append((current[0] - 1, current[1]))
                queue.append((current[0], current[1] + 1))
                queue.append((current[0], current[1] - 1))
        
        return distinct